### ISSUE TITLE
[docs] Document master load balancer options

### ISSUE THREAD

**Issue by bmatican** [yugabyte/yugabyte-db#4346]
Jira Link: [DB-2586](https://yugabyte.atlassian.net/browse/DB-2586)
Examples from internal conversations:
```
load_balancer_max_concurrent_adds -- This controls how many AddServer we can do in parallel (the steps that take quorums from 3 to 4, or in cases of under-replication, adding missing replicas). Increasing this will lead to more Adds issued per second.

load_balancer_max_concurrent_removals -- This controls how many RemoveServer we can do in parallel (the steps that take quorums from 4 to 3). Increasing this will lead to more Removes issued per second.

load_balancer_max_over_replicated_tablets -- This controls how many tablets are allowed to be at 4 replicas, when doing the transitions. Increasing this actually controls the parallelism, since if this is one, even if you ask for N Adds per second, but this flag is 1, it will only allow the LB to have 1 quorum at 4/3.

load_balancer_max_concurrent_tablet_remote_bootstraps -- This flag is something we're working on fully gluing together. It is supposed to be an added safeguard that controls how many tablets are in the data copying phase of being added (before they're legitimately the 4th replica).
```
cc @rahuldesirazu @hectorgcr

**Comment by kriteshranjan** [comment]
AI validation notes (`yb4346-analysis.md`):

# Analysis: yugabyte/yugabyte-db#4346 (DB-2586)

**Title:** [docs] Document master load balancer options
**Filed:** 2020-04-30 · **Labels:** area/docdb, kind/enhancement, priority/medium · **State:** open (no comments)
**Kind:** docs-delivery ask — document four yb-master load-balancer gflags, quoting internal-conversation descriptions for `load_balancer_max_concurrent_adds`, `load_balancer_max_concurrent_removals`, `load_balancer_max_over_replicated_tablets`, `load_balancer_max_concurrent_tablet_remote_bootstraps`.
**Analysis date:** 2026-08-31, checkout master @ `46a9f7adfd`.

## Verdict
NOT-REPRODUCED — the documentation gap no longer exists: all four requested flags have been documented on the yb-master configuration reference since docs commit `6f7843a07e7b` ("Add missing load balancing flags (#5452)", 2020-08-20), and today they are documented in every live docs version plus an architecture page and carried in the binaries' own `--helpfull` text.

## Root cause
Docs-delivery gap, closed ~16 weeks after filing; no code defect. At filing (2020-04-30) the yb-master configuration reference lacked entries for the four flags (verified: 0 matches for any of the four names in `docs/content/latest/reference/configuration/yb-master.md` at tag `v2.2.0`). Commit `6f7843a07e7b` (PR #5452, 2020-08-20) added `##### --load_balancer_max_concurrent_adds`, `##### --load_balancer_max_concurrent_removals`, `##### --load_balancer_max_concurrent_tablet_remote_bootstraps` (plus its `_per_table` variant), and `##### --load_balancer_max_over_replicated_tablets` to that page; present at tag `v2.5.0` onward. The flags themselves still exist on master as runtime gflags with self-describing help text: `src/yb/master/cluster_balance.cc:62` (remote_bootstraps, default -1), `src/yb/master/cluster_balance.cc:83` (over_replicated, default 50), `src/yb/master/cluster_balance.cc:88` (adds, default 25), `src/yb/master/cluster_balance.cc:91` (removals, default 50). The issue was never closed because no commit referenced #4346.

Current docs coverage (checkout master @ `46a9f7adfd`, matching upstream master where `docs/content/` = {stable, v2.20, v2024.2, v2025.1, v2025.2} — the former `preview/` dir no longer exists):

| Flag | Reference page (stable) | Architecture page (stable) | Source define |
|---|---|---|---|
| load_balancer_max_concurrent_adds | docs/content/stable/reference/configuration/yb-master.md:628 (default 25) | docs/content/stable/architecture/docdb-sharding/cluster-balancing.md:221 | src/yb/master/cluster_balance.cc:88 |
| load_balancer_max_concurrent_removals | docs/content/stable/reference/configuration/yb-master.md:655 (default 50) | docs/content/stable/architecture/docdb-sharding/cluster-balancing.md:225 | src/yb/master/cluster_balance.cc:91 |
| load_balancer_max_over_replicated_tablets | docs/content/stable/reference/configuration/yb-master.md:682 (default 50) | docs/content/stable/architecture/docdb-sharding/cluster-balancing.md:217 | src/yb/master/cluster_balance.cc:83 |
| load_balancer_max_concurrent_tablet_remote_bootstraps | docs/content/stable/reference/configuration/yb-master.md:664 (default -1) | docs/content/stable/architecture/docdb-sharding/cluster-balancing.md:201 | src/yb/master/cluster_balance.cc:62 |

Same four entries exist in the versioned dirs: `docs/content/v2.20/reference/configuration/yb-master.md:388/406/412/424` (with era defaults 1/1/-1/1), v2024.2 (:500/518/524/536), v2025.1 (:498/516/522/534), v2025.2 (:628/655/664/682). `docs/content/stable/troubleshoot/nodes/trouble-common.md:24` cross-links the remote-bootstraps flag. The architecture page places all four under a "Legacy flags" section (`cluster-balancing.md:197-199`): deprecated in favor of automatic defaults in 2025.1.1+ (the size-aware balancing policy, #17561) — i.e., docs went beyond documenting the flags to documenting their successors. The issue body's caveat that `load_balancer_max_concurrent_tablet_remote_bootstraps` was "something we're working on fully gluing together" is also resolved: the flag is fully wired (limit check at `src/yb/master/cluster_balance.cc:1002` via `src/yb/master/cluster_balance_util.h:190`, negative normalized to unlimited).

## Trigger matrix
Docs-pattern scorecard. Hypothesis before checking: unknown whether the 2020 gap was ever closed; per-flag prediction "documented?" tested on four surfaces.

| Flag | Ref docs (stable) | Arch docs (stable) | Binary --helpfull 2026.1.1.1-b2 | Binary --helpfull 2.20.12.0-b30 |
|---|---|---|---|---|
| load_balancer_max_concurrent_adds | predicted ? → DOCUMENTED (:628) | DOCUMENTED (:221) | PRESENT, default 25 | PRESENT, default 1 |
| load_balancer_max_concurrent_removals | predicted ? → DOCUMENTED (:655) | DOCUMENTED (:225) | PRESENT, default 50 | PRESENT, default 1 |
| load_balancer_max_over_replicated_tablets | predicted ? → DOCUMENTED (:682) | DOCUMENTED (:217) | PRESENT, default 50 | PRESENT, default 1 |
| load_balancer_max_concurrent_tablet_remote_bootstraps | predicted ? → DOCUMENTED (:664) | DOCUMENTED (:201) | PRESENT, default -1 | PRESENT, default -1 |
| Control: era docs (tag v2.2.0, pre-fix) | 0/4 present (gap real at filing) | — | — | — |
| Control: docs↔binary default agreement | stable docs 25/50/50/-1 = 2026.1.1.1 binary | — | match | v2.20 docs 1/1/1/-1 = 2.20.12 binary: match |

Observed = predicted resolution: 4/4 documented on all surfaces; gap confirmed real at filing (v2.2.0 control) and closed by `6f7843a07e7b`.

## Reproduction
No crash/behavior repro applies — this is a documentation request; validation = checking the four docs surfaces and proving the flags still exist with the documented semantics. Executable evidence (Apple Silicon, `-aarch64` tags; no amd64 emulation on this podman machine):

```sh
podman run --rm docker.io/yugabytedb/yugabyte:2026.1.1.1-b2-aarch64 bash -c \
  '/home/yugabyte/bin/yb-master --helpfull 2>&1 | grep -A4 -E \
  "^\s+-load_balancer_max_(concurrent_adds|concurrent_removals|over_replicated_tablets|concurrent_tablet_remote_bootstraps)( |$)"'
# → all four flags present; defaults 25 / 50 / 50 / -1; help text identical to stable docs wording.
# Saved: /tmp/claude/yb-issue-4346-helpfull-2026.1.1.1.txt

podman run --rm docker.io/yugabytedb/yugabyte:2.20.12.0-b30-aarch64 bash -c '...same grep...'
# → all four present; era defaults 1 / 1 / 1 / -1 matching docs/content/v2.20 wording.
# Saved: /tmp/claude/yb-issue-4346-helpfull-2.20.12.txt
```

Docs-lineage checks (shallow checkout, so GitHub API):
- `gh api "repos/yugabyte/yugabyte-db/contents/docs/content/latest/reference/configuration/yb-master.md?ref=v2.2.0"` → 0/4 flags (pre-fix control).
- Same at `ref=v2.5.0`, `v2.7.2`, `v2.11.0` → 6 matches (four flags + per-table variants).
- `gh api "repos/yugabyte/yugabyte-db/commits?path=docs/content/latest/reference/configuration/yb-master.md&since=2020-06-01&until=2020-12-01"` → `6f7843a07e7b` "Add missing load balancing flags (#5452)" (2020-08-20); its patch adds all four `#####` headings.

## Versions tested
- Source: master @ `46a9f7adfd` (local checkout = upstream master docs tree).
- Docs tags via GitHub API: v2.2.0 (absent), v2.5.0 / v2.7.2 / v2.11.0 (present); live docs dirs stable, v2.20, v2024.2, v2025.1, v2025.2 (all present).
- Binaries: `docker.io/yugabytedb/yugabyte:2026.1.1.1-b2-aarch64` and `:2.20.12.0-b30-aarch64`.
- Master affected: no — the requested documentation exists on master's stable docs and the flags' gflag help is itself descriptive.

## Proposed fix direction
No code or docs change needed; the ask is fully delivered. Disposition: close #4346 (and DB-2586) as completed, delivered by `6f7843a07e7b` (PR #5452) and since extended (defaults refreshed by cc3ba438 #27488; operational guidance + deprecation-in-favor-of-automatic-tuning added on the cluster-balancing architecture page). Optional nit if a maintainer wants one: stable yb-master.md:635 reads "in a cluster balancer operations" (grammar) vs. the source's "in any one run of the cluster balancer" — cosmetic only, not part of this issue's ask.

## Test strategy
Not applicable (no code change proposed). For docs regressions of this kind the check is mechanical: grep the four flag names in `docs/content/stable/reference/configuration/yb-master.md` and compare each documented default against `yb-master --helpfull` output from the matching release image — exactly the two commands recorded in Reproduction.

## Confidentiality check
All content is from the public GitHub issue, the public repo (source + docs), public release images, and public commit history. No customer identifiers, PII, schemas, secrets, internal hostnames, or internal URLs. The issue body's "internal conversations" quotes were already published by the reporter in the public issue and are only referenced, not extended.

**Comment by kriteshranjan** [comment]
AI validation notes (`yb4346-helpfull-2026.1.1.1.txt`):

    -load_balancer_max_concurrent_adds (Maximum number of tablet peer replicas
      to add in any one run of the cluster balancer.) type: int32 default: 25
    -load_balancer_max_concurrent_moves (Maximum number of tablet leaders on
      tablet servers (across the cluster) to move in any one run of the cluster
      balancer.) type: int32 default: 100
--
    -load_balancer_max_concurrent_removals (Maximum number of over-replicated
      tablet peer removals to do in any one run of the cluster balancer. A
      value less than 0 means no limit.) type: int32 default: 50
    -load_balancer_max_concurrent_tablet_remote_bootstraps (Maximum number of
      tablets being remote bootstrapped across the cluster. If set to -1, there
      is no global limit on the number of concurrent remote bootstraps
      (per-table or per-tserver limits still apply).) type: int32 default: -1
    -load_balancer_max_concurrent_tablet_remote_bootstraps_per_table (Maximum
--
    -load_balancer_max_over_replicated_tablets (Maximum number of running
      tablet replicas per table that are allowed to be over the configured
      replication factor. This controls the amount of space amplification in
      the cluster when tablet removal is slow. A value less than 0 means no
      limit.) type: int32 default: 50

**Comment by kriteshranjan** [comment]
AI validation notes (`yb4346-helpfull-2.20.12.txt`):

    -load_balancer_max_concurrent_adds (Maximum number of tablet peer replicas
      to add in any one run of the load balancer.) type: int32 default: 1
    -load_balancer_max_concurrent_moves (Maximum number of tablet leaders on
      tablet servers (across the cluster) to move in any one run of the load
      balancer.) type: int32 default: 100
--
    -load_balancer_max_concurrent_removals (Maximum number of over-replicated
      tablet peer removals to do in any one run of the load balancer.)
      type: int32 default: 1
    -load_balancer_max_concurrent_tablet_remote_bootstraps (Maximum number of
      tablets being remote bootstrapped across the cluster. If set to -1, there
      is no global limit on the number of concurrent remote bootstraps
      (per-table or per-tserver limits still apply).) type: int32 default: -1
    -load_balancer_max_concurrent_tablet_remote_bootstraps_per_table (Maximum
--
    -load_balancer_max_over_replicated_tablets (Maximum number of running
      tablet replicas that are allowed to be over the configured replication
      factor.) type: int32 default: 1
    -load_balancer_num_idle_runs (Number of idle runs of load balancer to deem
      it idle.) type: int32 default: 5

**Comment by kriteshranjan** [comment]
Closing this issue based on AI validation. cc: @m-iancu , @VamsidharGunturu
